proposals/0000-android-12-issues.md

```diff
@@ -0,0 +1,353 @@
+
+~~~ java
+
+int flag = android.os.Build.VERSION.SDK_INT >= Build.VERSION_CODES.S ? PendingIntent.FLAG_IMMUTABLE : 0;
+PendingIntent pendingIntent = PendingIntent.getActivity(this, 0, intent, flag);
+
+~~~
+
+### Service Notification Delays
+
+Starting in Android 12 when a service tries to present a notification, Android may delay showing the notification for up to 10 seconds. This is to try to allow the service to complete before the notification is presented. If we have connected to a system before we may want to present the RouterService notifications immediately. App developers may also want to display the notifications related to the `SdlService` immediately. This can be achieved by setting the foregroundServiceBehavior flag to `Notification.FOREGROUND_SERVICE_IMMEDIATE`.
+
+~~~ java
+
+private void safeStartForeground(int id, Notification notification) {
+    try {
+        if (notification == null) {
+            if (hasConnectedBefore && android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.S) {
+                Notification.Builder builder =
+                        new Notification.Builder(this, SDL_NOTIFICATION_CHANNEL_ID)
+                            .setContentTitle("SmartDeviceLink")
+                            .setContentText("Service Running")
+                            .setForegroundServiceBehavior(Notification.FOREGROUND_SERVICE_IMMEDIATE);
+                notification = builder.build();
+            } else {
+                //Try the NotificationCompat this time in case there was a previous error
+                NotificationCompat.Builder builder =
+                        new NotificationCompat.Builder(this, SDL_NOTIFICATION_CHANNEL_ID)
+                                .setContentTitle("SmartDeviceLink")
+                                .setContentText("Service Running");
+
+                notification = builder.build();
+            }
+        }
+        startForeground(id, notification);
+        DebugTool.logInfo(TAG, "Entered the foreground - " + System.currentTimeMillis());
+    } catch (Exception e) {
+        DebugTool.logError(TAG, "Unable to start service in foreground", e);
+    }
+}
+
+~~~
+
+## Potential downsides
+
+### Foreground Services
+
+With the proposed solution of using PendingIntents to start the apps `SdlService` from the RouterService this creates a potential security risk as the apps `SdlService` will now be required to have the `android:exported` attribute set to true. This could expose apps to have their `SdlService` be started by apps that are not SDL certified. We could also create a custom SDL `<permission>` to be used by apps as a requirement to be able to start the `SdlService` but nothing is stoping developers from listing that custom permission in their `AndroidManifest.xml`.   
```

Comment:
I think the custom permission solution should be kept in the "Alternative solutions"

Reply:
```suggestion
Using `PendingIntents` to start the apps `SdlService` from the RouterService creates a potential security risk as the apps `SdlService` will now be required to have the `android:exported` attribute set to true. This could expose apps to have their `SdlService` be started by apps that are not SDL certified. We could also create a custom SDL `<permission>` to be used by apps as a requirement to be able to start the `SdlService` but nothing is stoping developers from listing that custom permission in their `AndroidManifest.xml`.   
```